File: 3/Lua/3.lua
local large = 600851475143
function isPrime(value)
	for i=2, math.floor(value/2)-1 do
		if value % i == 0 then
			return false
		end
	end
	return true
end
for i=math.floor(math.sqrt(large))+1, 2, -1 do
	if large % i == 0 and isPrime(i) then
		print(i)
		return
	end
end

File: 3/Haskell/3.hs
-- Problem 3
-- The prime factors of 13195 are 5, 7, 13 and 29.
-- What is the largest prime factor of the number 600851475143 ?

isDivisible :: (Integral a) => a -> a -> Bool
isDivisible numerator denominator = (mod numerator denominator) == 0

firstPrimeTest :: (Integral a) => a -> a
firstPrimeTest primeCandidate = ceiling $ sqrt $ fromIntegral primeCandidate

isPrime :: (Integral a) => a -> Bool
isPrime primeCandidate = not (any (isDivisible primeCandidate) [2..firstPrimeTest primeCandidate])

main :: IO ()
main =
  putStrLn $ show $
    head $
    filter isPrime $
    filter (isDivisible 600851475143)
    [firstPrimeTest 600851475143, (firstPrimeTest 600851475143)-1..1]
